_posts/2020-07-31-fnn-vae-for-noisy-timeseries/fnn_vae_for_noisy_timeseries.Rmd

```diff
@@ -0,0 +1,509 @@
+interested reader, to keep the post at reasonable length.)
+
+One last contextual thing, in case you haven't read the two previous posts and would like to jump in here directly. We're
+doing time series forecasting; so why this talk of autoencoders? Shouldn't we just be comparing an LSTM (or some other type of
+RNN, for that matter) to a convnet? In fact, the necessity of a latent representation is due to the very idea of FNN: The
+latent code is supposed to reflect the true attractor of a dynamical system. That is, if the attractor of the underlying
+system is roughly two-dimensional, we hope to find that just two of the latent variables have considerable variance. (This
+reasoning is explained in a lot of detail in the previous posts.)
+
+## FNN-VAE
+
+So, let's start with the code for our new model.
+
+The encoder takes the time series, of format `batch_size x num_timesteps x num_features` just like in the LSTM case, and
+produces a flat, 10-dimensional output: the latent code, which FNN loss is computed on.
+
+```{r, eval = FALSE}
+
+library(tensorflow)
+library(keras)
+library(tfdatasets)
+library(tfautograph)
+library(reticulate)
+library(purrr)
+
+vae_encoder_model <- function(n_timesteps,
+                               n_features,
+                               n_latent,
+                               name = NULL) {
+  keras_model_custom(name = name, function(self) {
+    self$conv1 <- layer_conv_1d(kernel_size = 3,
+                                filters = 16,
+                                strides = 2)
+    self$act1 <- layer_activation_leaky_relu()
+    self$batchnorm1 <- layer_batch_normalization()
+    self$conv2 <- layer_conv_1d(kernel_size = 7,
+                                filters = 32,
+                                strides = 2)
+    self$act2 <- layer_activation_leaky_relu()
+    self$batchnorm2 <- layer_batch_normalization()
+    self$conv3 <- layer_conv_1d(kernel_size = 9,
+                                filters = 64,
+                                strides = 2)
+    self$act3 <- layer_activation_leaky_relu()
+    self$batchnorm3 <- layer_batch_normalization()
+    self$conv4 <- layer_conv_1d(
+      kernel_size = 9,
+      filters = n_latent,
+      strides = 2,
+      activation = "linear" 
+    )
+    self$batchnorm4 <- layer_batch_normalization()
+    self$flat <- layer_flatten()
+    
+    function (x, mask = NULL) {
+      x %>%
+        self$conv1() %>%
+        self$act1() %>%
+        self$batchnorm1() %>%
+        self$conv2() %>%
+        self$act2() %>%
+        self$batchnorm2() %>%
+        self$conv3() %>%
+        self$act3() %>%
+        self$batchnorm3() %>%
+        self$conv4() %>%
+        self$batchnorm4() %>%
+        self$flat()
+    }
+  })
+}
+
+```
+
+The decoder starts from this -- flat -- representation and decompresses it into a time sequence. In both encoder and decoder
```

Comment:
I think you're using the dashes here for emphasis. You might consider bold or italics instead. Feel free to ignore, as always. 🤗

Reply:
yeah you know, I think I really want the dashes, not <em></em> so to say ... 
What you said about "visual" before, that might explain it - for me, I hear a spoken sentence, and there's a pause before and after "flat"